async support v2

999eagle opened the pull request:
As discussed in #314, this is a new attempt to add support for `async` reading.

This PR is still WIP and open for discussion on implementation details.

Mingun commented:
Response to https://github.com/tafia/quick-xml/pull/314#issuecomment-1180124604

Actually, I'll plan to merge similar traits `XmlRead` and `XmlSource` into one and use approach with a special intermediate struct `SliceReader` / `IoReader` instead of directly implementing `XmlSource` for `BufRead` and `&[u8]`. With that approach it should be possible to have an `AsyncSliceReader` / `AsyncIoReader` (which would wrap only a inner `reader`) for which new traits will be implemented.

So it is good time for you to investigate that surface and how it can be integrated with `async` support

999eagle commented:
Okay, I've moved a lot of stuff around to hopefully separate some things better. I've also implemented async support on the existing `Reader`. The following things have been changed as of now:

- The `Reader` is now generic over the parser in use. This separates namespaced access and non-namespaced access through type checking at compile time.
- To implement this, there is now a `Parser`-trait, implemented by a `DefaultParser` and a `NamespacedParser`.
- The `Reader` is now built via a builder-pattern so no changes to the parser settings (like whether whitespace is trimmed or comments/end tags are checked) are possible after being built. This does change the public API and thus is a breaking change but allows the `Parser` separation and thus eliminates a lot of duplicate code.
- There is now an `AsyncXmlSource` trait that's automatically implemented for everything that implements `tokio`'s `AsyncBufRead` trait.
- The `Reader` now has some `read_..._async` methods. Using the same names for sync and async methods sadly won't work  as that would require to tell Rust that no type can ever implement both `XmlSource` and `AsyncXmlSource` (or `BufRead` and `AsyncBufRead`) at the same time which is (for now) impossible.
- There is a new `test_async` test which is rather bare-bones as of now but I'll work on that.
- Currently, tests with the `serialize` feature are broken because I haven't changed  the deserialization to work with the new `Reader` yet.

Mingun commented:
1852c70d75fce7831062f22c068379d4b89016ed is a pretty big change which is hard to review. Could you split this commit into several? At least I can suggest extract introducing builder pattern into a separate commit. Also, can you rebase your work and squash up all fixup changes as much as possible before the finishing work (for example, you've changed `AsyncXmlSource` trait in the commit where you are actually implement it. Actually, it would be better to introduce this trait in the same commit where you implement it).

Also, some changes seems to a little complicated to me. Actually I was thinking about something like (signatures approximate, just to illustrate the idea):
```rust
struct SliceReader<'i>(&'i [u8]);
struct IoReader<R: BufRead>(R);
struct AsyncIoReader<R: AsyncBufRead>(R);

/// The existing Reader struct
pub struct Reader<R> {
  reader: R,
  // ...
}

impl<R> Reader<R> {
  // common parsing code, setup code
}

impl<'i> Reader<SliceReader<'i>> {
  pub fn read_event(&mut self) { /* ... */ }
  pub fn read_event_async(&mut self) { /* ... */ }
}

impl<R: BufRead> Reader<IoReader<R>> {
  pub fn read_event_into(&mut self, buf: &mut Vec<u8>) { /* ... */ }
}

impl<R: AsyncBufRead> Reader<AsyncIoReader<R>> {
  pub fn read_event_into_async(&mut self, buf: &mut Vec<u8>) { /* ... */ }
}

////////////////////////////////////////////////////////////////////////////

/// The struct that I would introduce in my namespaces PR
pub struct NsReader<R> {
  reader: Reader<R>,
  // ...
}

impl<R> NsReader<R> {
  // common parsing code, setup code
}

impl<'i> NsReader<SliceReader<'i>> {
  pub fn read_event(&mut self) { /* ... */ }
  pub fn read_event_async(&mut self) { /* ... */ }
}

impl<R: BufRead> NsReader<IoReader<R>> {
  pub fn read_event_into(&mut self, buf: &mut Vec<u8>) { /* ... */ }
}

impl<R: AsyncBufRead> NsReader<AsyncIoReader<R>> {
  pub fn read_event_into_async(&mut self, buf: &mut Vec<u8>) { /* ... */ }
}
```

999eagle commented:
Okay, so you'd prefer having the `NsReader<R>` wrap the "standard" `Reader<R>` and then implement the basic parsing functions like that. I think that would work too for async support. Your example code also greatly simplifies reading as the generic bound for the buffer in `XmlSource` doesn't exist anymore and the different wrapper structs allow for strictly separating sync/async access.
With that approach, I think it's better to abandon my current code entirely and start anew. I'll try again from the current `master` using your approach.

999eagle commented:
@Mingun I think I've reimplemented this now the way you suggested and with minimal unrelated changes. The builder pattern can still be added again in a later PR. I also didn't separate the namespaced read methods in this PR but it should be relatively easy to add.

Mingun reviewed (changes requested):
This is a big change, so now I made only a quick overview. I'll plan to made a second review, probably after my remarks will be addressed.

Fist of all, this variant looks much more simpler and better that a previous attempt. For the maintenance purposes I prefer that each commit was as smaller as possible and focusing on the one thing. Your f9cdf2796c034c0a65c439b09bdc5f14920c5911 commit actually does the much more that just splitting reader. I suggest to split it for a several commits:
- the one that changes usage of buffered methods in tests and examples into borrowing ones
- the one that introduces a helper macros which conditionally includes `await` and `async` keywords
- the one that moves code to the separate files
- the one that actually implements `async` reader. The adding tests commit then probably can be included in that commit
- any other that you'll decide to made to keep the review burden low

Please ensure, that after each commit code in the buildable state by running commands that CI run and additionally `cargo doc --all-features`. If some tests will temporary fail, please include the list of failed tests to the commit message (just copy output of `cargo test` after "failures:")

Try to keep copies of code low. Use helper functions / traits / macros if needed. For example, the original `read_bang_element` probably contains a bug, because `read` are always `>=1` but there is `read == 0` check, I do not want to copy it over the whole project.

Also add an entry to the Changelog.md.

It also will be a great to run benchmarks for the async code and compare results with the sync code, so it will be a plus if you add these benchmarks.

@dralley, I also would to hear your opinion

999eagle commented:
Okay, I think I have addressed all your concerns now. Feel free to mark your discussions as resolved where it applies and keep them open otherwise. I've added some docs, updated examples and split the commits as best as I could.

999eagle commented:
Rebased onto current master to resolve conflicts.

dralley commented:
@999eagle Apologies for all the breakages!

If there is anything in this PR which is independent of the fundamental changes, such as perhaps commits #1 and #4, or test updates in which don't strictly involve the async stuff, it would be fantastic if you could split it off into a smaller PR which we can review and merge quickly.

999eagle commented:
Hi @dralley, while those two commits are somewhat independent of the other changes, they don't really do anything useful on their own. The first commit is nothing but a requirement for testing the changes of the third commit, which in turn requires an example for explicitly buffered access for files as most examples just read strings or byte slices and thus don't have buffered reading anymore.
In my opinion, it'd make sense to split this PR into two separate PRs at most: Splitting the `Reader` into the explicit `SliceReader` and `IoReader` (that is, commits 1 through 4 from this current PR) for the first PR and all the `async`-specific implementation in a second PR. If that's what you want to do, tell me and I'll open a new PR for the non-async changes.

dralley commented:
@999eagle Yes, that makes perfect sense, and it would be helpful.

dralley commented on src/reader.rs:
```diff
@@ -123,9 +141,8 @@ impl EncodingRef {
-/// let mut buf = Vec::new();
 /// loop {
-///     match reader.read_event_into(&mut buf) {
+///     match reader.read_event() {
```
Agreed with this, the README does have an example, but it would be helpful for the documentation to have one readily available as well.

dralley commented on src/reader/slice_reader.rs:
```diff
@@ -0,0 +1,523 @@
+    }
+}
+
+/// Private functions for a [`Reader`] based on an [`IoReader`].
```
Typo I believe?

```suggestion
/// Private functions for a [`Reader`] based on a[`SliceReader`].
```

dralley commented:
Almost done reviewing the first 4 commits (which as you said will be split into the new PR), and it is looking fine so far.  I really appreciate all the work you've put in to get this cleaned up.

999eagle commented:
@dralley I've opened #425 for the first few changes. I've also rebased onto the current master. As this PR is now based on #425, it must be merged afterwards.

dralley commented:
This is relevant: https://blog.rust-lang.org/inside-rust/2022/07/27/keyword-generics.html

Anything that comes of that initiative would be months or years down the road, but it might be worthwhile to provide feedback and examples early in the process.
